webdriver-spec.html

```diff
@@ -5148,35 +5148,65 @@
    <li><p>Remove all <var>undo actions</var> from the <a>current
     session</a>'s <a>input cancel list</a>.
 
-  <li><p>Let <var>i</var> be 0.
+   <li><p>Let <var>keyboard</var> be the first found <a>key input
+    source</a> in the <a>current session</a>'s <a>input state</a>
+    or <code>undefined</code> if no <a>key input source</a> can be
+    found.
 
-  <li><p>Let <var>length</var> be the length of <var>character array</var>.
+   <li><p>If <var>keyboard</var> is <code>undefined</code> <a>create a
+    new key input state object</a> and add it to the <a>current
+    session</a>'s <a>input state</a>. Let <var>keyboard</var> equal
+    this newly created <a>key input state</a>.
 
-  <li><p>While <var>i</var> &lt; <var>length</var>:
+   <li><p>Let <var>undo actions</var> be an empty list.
 
-    <ol>
-      <li><p>Let <var>char</var> be the entry in <var>character array</var> at
-        index <var>i</var>.
-
-      <li><p>Let <var>key to type</var> be the result of <a>processing keys</a>
-        with argument <var>key</var>.
-
-      <li><p>If the <var>key to type</var> is a <a>modifier key</a> then set the relevant
-        attribute on <a>Keyboard Event</a> to true for the rest of this loop.
-
-      <li><p>Generate <a data-lt=keyDown-Event>keyDown</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
+   <li><p>For each <var>char</var> corresponding to an indexed property in <var>character array</var>.
 
-      <li><p>Generate <a data-lt=keyPress-Event>keyPress</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
-
-      <li><p>Generate <a data-lt=keyUp-Event>keyUp</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
+    <ol>
+     <li><p>Let <var>key to type</var> be the result of <a>processing keys</a> with argument <var>char</var>.
+
+     <li><p>Let <var>action</var> be an <a>action object</a>
+      constructed with
+      arguments <var>keyboard</var>'s <var>id</var>, <code>"key"</code>,
+      and <code>"keyDown"</code>.
+
+     <li><p>Set the <code>value</code> property of <var>action</var> to <var>key to type</var>.
+
+     <li><p><a>Dispatch a keyDown action</a> with
+     arguments <var>action</var> and <var>keyboard</var>.
```

Comment:
The second argument to `keyDown` is an `input state`, not a `device state`. Presumably you're in the middle of a series of changes to that terminology? In either case, `keyboard` has been defined as an `input state` so it should be fine to just pass it in.